### PR TITLE
CPU discovery

### PR DESCRIPTION
**Pull request by squall0gd**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: cluster/addons/iso-client/coreaffinity-isolator.go [hunk 1]
```diff
@@ -12,11 +12,12 @@ import (
 	//	"k8s.io/apimachinery/pkg/util/uuid"
 
 	aff "k8s.io/kubernetes/cluster/addons/iso-client/coreaffinity"
+	"k8s.io/kubernetes/cluster/addons/iso-client/discovery"
 )
 
 const (
 	// kubelet eventDispatcher address
-	eventDispatcherAddress = "localhost:5433"
+	eventDispatcherAddress = "localhost:6443"
```
Comment (PiotrProkop) [inline cluster/addons/iso-client/coreaffinity-isolator.go]: Change this port to 5433 .

File: cluster/addons/iso-client/coreaffinity-isolator.go [hunk 2]
```diff
@@ -27,10 +28,17 @@ const (
 func main() {
 	flag.Parse()
 	glog.Info("Starting ...")
+	cpuTopo, err := discovery.DiscoverTopology()
+	if err != nil {
+		glog.Fatalf("Cannot retrive CPU topology: %q", err)
```
Comment (PiotrProkop) [inline cluster/addons/iso-client/coreaffinity-isolator.go:33]: Add os.Exit(1)

File: cluster/addons/iso-client/cputopology/cpu.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	socketCounter := 0
+	for _, cpu := range c.CPU {
+		if cpu.SocketID > socketCounter {
+			socketCounter = cpu.SocketID
+		}
+	}
+	return socketCounter + 1
+}
+
+func (c *CPUTopology) GetSocket(socketID int) (cores []CPU) {
```
Comment (PiotrProkop) [inline cluster/addons/iso-client/cputopology/cpu.go:37]: I would create a Socket struct and return it .

File: cluster/addons/iso-client/cputopology/cpu.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	for _, cpu := range c.CPU {
+		if cpu.SocketID == socketID {
+			cores = append(cores, cpu)
+		}
+	}
+	return
+}
+
+func (c *CPUTopology) GetCore(socketID, coreID int) (cpus []CPU) {
```
Comment (PiotrProkop) [inline cluster/addons/iso-client/cputopology/cpu.go:46]: I would create a Core struct and return it.

File: cluster/addons/iso-client/cputopology/cpu.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	for _, cpu := range c.CPU {
+		if cpu.SocketID == socketID && cpu.CoreID == coreID {
+			cpus = append(cpus, cpu)
+		}
+	}
+	return
+}
+
+func (c *CPUTopology) GetCPU(cpuid int) CPU {
```
Comment (PiotrProkop) [inline cluster/addons/iso-client/cputopology/cpu.go]: Delete this function and rename getCpuPointer as GetCpu.

**Comment by squall0gd** [comment]
I'm not so sure TBH about value of additional `Socket` and `Core` struct. It could be much more complicated then current approach. Anyway, I agree that this two methods (`GetSocket` and `GetCore`) could have much more better names :)

**Comment by squall0gd** [comment]
I'm also not so sure what about hyperthreaded cpus. Should we recognize which cpu is hyperthreaded? /cc @ConnorDoyle

File: cluster/addons/iso-client/coreaffinity-isolator.go (hunk 2 above)
Comment (ConnorDoyle) [reply cluster/addons/iso-client/coreaffinity-isolator.go:33]: It's implicit in Fatalf

File: cluster/addons/iso-client/cputopology/cpu.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+type CPU struct {
+	// SocketID represents socket where CPU is.
+	SocketID int
+	// CoreID represents core where CPU is.
+	CoreID int
+	// CPUID is unique CPU number.
+	CPUID int
+	// IsIsolated stores information is CPU is not use by OS scheduler.
+	IsIsolated bool
+	// IsInUse stores information
```
Comment (ConnorDoyle) [inline cluster/addons/iso-client/cputopology/cpu.go]: Could you expand on this comment a bit?

File: cluster/addons/iso-client/cputopology/cpu.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+		}
+	}
+	return nil
+}
+
+func (c *CPUTopology) Reserve(cpuid int) {
+	c.Lock.Lock()
+	cpuAddress := c.getCPUPointer(cpuid)
+	cpuAddress.IsInUse = true
+	c.Lock.Unlock()
```
Comment (ConnorDoyle) [inline cluster/addons/iso-client/cputopology/cpu.go]: Style nit: defer the unlock immediately after acquiring

File: cluster/addons/iso-client/cputopology/cpu.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+func (c *CPUTopology) Reclaim(cpuid int) {
+	c.Lock.Lock()
+	cpuAddress := c.getCPUPointer(cpuid)
+	cpuAddress.IsInUse = false
+	c.Lock.Unlock()
```
Comment (ConnorDoyle) [inline cluster/addons/iso-client/cputopology/cpu.go]: Style nit: defer the unlock immediately after acquiring

File: cluster/addons/iso-client/cputopology/cpu.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+			return &c.CPU[idx]
+		}
+	}
+	return nil
+}
+
+func (c *CPUTopology) Reserve(cpuid int) {
+	c.Lock.Lock()
+	cpuAddress := c.getCPUPointer(cpuid)
+	cpuAddress.IsInUse = true
```
Comment (ConnorDoyle) [inline cluster/addons/iso-client/cputopology/cpu.go]: What happens if the CPU is already allocated?

File: cluster/addons/iso-client/discovery/cpu_discovery.go [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+func cpuParse(lscpuStringOutput string, isolatedCPUs []int) (*cputopology.CPUTopology, error) {
+	lscpuRawSlice := strings.Split(lscpuStringOutput, "\n")
+	var cpuTopology cputopology.CPUTopology
+
+	for _, line := range lscpuRawSlice {
+		if strings.Index(line, "#") != -1 || strings.Compare(line, "") == 0 {
+			continue
+		}
+		cpuinfo := strings.Split(line, ",")
```
Comment (ConnorDoyle) [inline cluster/addons/iso-client/discovery/cpu_discovery.go]: Why not fmt.Sscanf?

File: cluster/addons/iso-client/cputopology/cpu.go [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+	// IsInUse stores information
+	IsInUse bool
+}
+
+type CPUTopology struct {
+	CPU  []CPU
+	Lock sync.Mutex
+}
+
+func (c *CPUTopology) GetSocketNumber() int {
```
Comment (ConnorDoyle) [inline cluster/addons/iso-client/cputopology/cpu.go]: Naming nit: GetNumSockets?

**Review by ConnorDoyle — COMMENTED** [review]
Regarding HT cores, ultimately that decision will depend on each isolator strategy. Since this is a PoC we should just do enough to illustrate a sample strategy without confusing the issue. For example: enumerate the isolated CPUs and advertise one OIR for each isolated physical core. In general HT cores will be a useful part of the topology. If the intent here is to make a reusable topo library then it should support those kinds of queries eventually.

File: cluster/addons/iso-client/cputopology/cpu.go [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+	// SocketID represents socket where CPU is.
+	SocketID int
+	// CoreID represents core where CPU is.
+	CoreID int
+	// CPUID is unique CPU number.
+	CPUID int
+	// IsIsolated stores information is CPU is not use by OS scheduler.
+	IsIsolated bool
+	// IsInUse stores information
+	IsInUse bool
```
Comment (ConnorDoyle) [inline cluster/addons/iso-client/cputopology/cpu.go]: An alternative to storing this here could be to maintain a collection of Binding (pod => []CPU) in the isolator.

Comment (ConnorDoyle) [reply cluster/addons/iso-client/cputopology/cpu.go]: That would eliminate the lock in this struct.

**Review by ConnorDoyle — APPROVED** [review]
Is it worth adding a unit test or two?

**Comment by squall0gd** [comment]
@ConnorDoyle I think so :)

**Pull request merged**